docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+   `ClientState` and `ConsensusState` structs themselves are theoretically
+   possible by providing a path to translate an older ClientState struct into
+   the new ClientState struct; however this is not currently implemented.
+  
+#### Upgrade Process Step-by-step
+
+An IBC-connected Tendermint chain will take the following steps to completely
+upgrade its own chain and counterparty's IBC client:
+
+1. Upgrade chain through governance
+   1. Create a 02-client
+      [UpgradeProposal](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/proto-docs.md#upgradeproposal)
+      with an `UpgradePlan` and a new IBC `ClientState` in the
+      `UpgradedClientState` field with the following remarks:
+        * The `UpgradePlan` must specify an upgrade height only (no upgrade
+          time)
+        * The `ClientState` should only include the fields common to all valid
+          clients and zero out any client-customizable fields (such as
+          `TrustingPeriod`).
+   2. Vote on and pass the `UpgradeProposal`
+   3. Commit `UpgradedClient` by the upgrade module under the following key:
+
+        ```md
+        upgrade/UpgradedIBCState/{upgradeHeight}/upgradedClient
+        ```
+
+   4. Commit an initial consensus state by upgrade module on the block right
+      before the upgrade height for the next chain under the following key:
+
+        ```md
+        upgrade/UpgradedIBCState/{upgradeHeight}/upgradedConsState
+        ```
+
+2. Submit an upgrade client message by relayers to the counterparty chain
+   1. Wait for the upgrading chain to reach the upgrade height and halt
+   2. Query a full node for the proofs of `UpgradedClient` and
+      `UpgradedConsensusState` at the last height of the old chain
+   3. Update the counterparty client to the last height of the old chain using
+      the `UpdateClient` msg
+   4. Submit an `UpgradeClient` msg to the counterparty chain with the
+      `UpgradedClient`, `UpgradedConsensusState` and their respective proofs
+
+3. Process the upgrade message on the counterparty chain</br> IBC handler upon
+receiving a `MsgUpgradeClient` message performs basic validations (BV),
+client-specific validations (SV) and lastly execution (E) steps as follows:
+   1. (BV) Check that the current client is not frozen
+   2. (BV) Check if the latest consensus state is within the trust period
+   3. (BV) Check if the message containing proofs decoded successfully
+   4. (SV) Verify that the upgradedClient be of a Tendermint `ClientState` type
+   5. (SV) Verify that the upgradedConsensusState be of a Tendermint
+      `ConsensusState` type
+   6. (SV) Check the latest height of the current client state does not have the
+      same revision number or has a greater height than the committed client
+   7. (SV) Match any Tendermint chain specified parameter in upgraded client
+      such as ChainID, UnbondingPeriod, and ProofSpecs with the committed client
```

Comment:
Can you point to me where this is in ibc-go? I don't see it anywhere

Reply:
Check [this](https://github.com/cosmos/ibc-go/blob/c86d27fc280cfb342a9e4689b381e5823441b694/modules/light-clients/07-tendermint/upgrade.go#L25) out. It was the place I wanna use that mentioned `validate()` method in [#384](https://github.com/cosmos/ibc-rs/issues/384)

Reply:
> Check [this](https://github.com/cosmos/ibc-go/blob/c86d27fc280cfb342a9e4689b381e5823441b694/modules/light-clients/07-tendermint/upgrade.go#L25) out.

Sorry, I meant the code that actually checks that rather than the "written statement". The only one I was able to find was [the `chainID` check](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/client_state.go#L129).

> It was the place I wanna use that mentioned validate() method in #384 

Notice that `validate()` [is called](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/upgrade.go#L101-L109) right after the constructor, so I don't see the gain in separating the validation from the constructor. And I still believe it is a code smell to allow an object to be constructed in an invalid state.

Reply:
>Sorry, I meant the code that actually checks that rather than the "written statement". The only one I was able to find was [the chainID check](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/client_state.go#L129).

Right, It is stated but the only check, on `ibc-go`, is done for the `ChainID` plus the validation of the constructed fields of the upgraded client state....
I added a check for `unboundingPeriod`, which does not exist on `ibc-go` (as you also pointed out), but for others I think it could be restrictive if something be placed. Though, I am still investigating.

>Notice that validate() [is called](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/upgrade.go#L101-L109) right after the constructor, so I don't see the gain in separating the validation from the constructor. And I still believe it is a code smell to allow an object to be constructed in an invalid state.

Actually, the main concern has been elaborated [here](https://github.com/cosmos/ibc-rs/issues/384#issuecomment-1407034388), but maybe we should leave it as is.

Reply:
Let's move the `validate()` conversation to #384